// @metaplex-foundation/mpl-candy-machine/dist/src/generated/instructions/setCollectionDuringMint.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSetCollectionDuringMintInstruction = void 0;
const beet = __importStar(require("@metaplex-foundation/beet"));
const web3 = __importStar(require("@solana/web3.js"));
const setCollectionDuringMintStruct = new beet.BeetArgsStruct([['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)]], 'SetCollectionDuringMintInstructionArgs');
const setCollectionDuringMintInstructionDiscriminator = [103, 17, 200, 25, 118, 95, 125, 61];
function createSetCollectionDuringMintInstruction(accounts) {
    const { candyMachine, metadata, payer, collectionPda, tokenMetadataProgram, instructions, collectionMint, collectionMetadata, collectionMasterEdition, authority, collectionAuthorityRecord, } = accounts;
    const [data] = setCollectionDuringMintStruct.serialize({
        instructionDiscriminator: setCollectionDuringMintInstructionDiscriminator,
    });
    const keys = [
        {
            pubkey: candyMachine,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: metadata,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: payer,
            isWritable: false,
            isSigner: true,
        },
        {
            pubkey: collectionPda,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: tokenMetadataProgram,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: instructions,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: collectionMint,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: collectionMetadata,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: collectionMasterEdition,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: authority,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: collectionAuthorityRecord,
            isWritable: false,
            isSigner: false,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId: new web3.PublicKey('cndy3Z4yapfJBmL3ShUp5exZKqR3z33thTzeNMm2gRZ'),
        keys,
        data,
    });
    return ix;
}
exports.createSetCollectionDuringMintInstruction = createSetCollectionDuringMintInstruction;
