// @metaplex-foundation/mpl-candy-machine/dist/src/generated/errors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.errorFromName = exports.errorFromCode = exports.MismatchedCollectionMintError = exports.MismatchedCollectionPDAError = exports.IncorrectCollectionAuthorityError = exports.IncorrectSlotHashesPubkeyError = exports.CannotSwitchToHiddenSettingsError = exports.SuspiciousTransactionError = exports.InvalidStringError = exports.CannotFindUsableConfigLineError = exports.NetworkExpireFeatureMissingError = exports.GatewayTokenExpireTimeInvalidError = exports.GatewayTokenMissingError = exports.GatewayAppMissingError = exports.TokenBurnFailedError = exports.NoWhitelistTokenError = exports.PublicKeyMismatchError = exports.DerivedKeyInvalidError = exports.CannotChangeNumberOfLinesError = exports.HiddenSettingsConfigsDoNotHaveConfigLinesError = exports.CandyMachineNotLiveError = exports.CandyMachineEmptyError = exports.TokenTransferFailedError = exports.NotEnoughSOLError = exports.NotEnoughTokensError = exports.UuidMustBeExactly6LengthError = exports.TooManyCreatorsError = exports.NumericalOverflowErrorError = exports.IndexGreaterThanLengthError = exports.MintMismatchError = exports.UninitializedError = exports.IncorrectOwnerError = void 0;
const createErrorFromCodeLookup = new Map();
const createErrorFromNameLookup = new Map();
class IncorrectOwnerError extends Error {
    constructor() {
        super('Account does not have correct owner!');
        this.code = 0x1770;
        this.name = 'IncorrectOwner';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, IncorrectOwnerError);
        }
    }
}
exports.IncorrectOwnerError = IncorrectOwnerError;
createErrorFromCodeLookup.set(0x1770, () => new IncorrectOwnerError());
createErrorFromNameLookup.set('IncorrectOwner', () => new IncorrectOwnerError());
class UninitializedError extends Error {
    constructor() {
        super('Account is not initialized!');
        this.code = 0x1771;
        this.name = 'Uninitialized';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, UninitializedError);
        }
    }
}
exports.UninitializedError = UninitializedError;
createErrorFromCodeLookup.set(0x1771, () => new UninitializedError());
createErrorFromNameLookup.set('Uninitialized', () => new UninitializedError());
class MintMismatchError extends Error {
    constructor() {
        super('Mint Mismatch!');
        this.code = 0x1772;
        this.name = 'MintMismatch';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, MintMismatchError);
        }
    }
}
exports.MintMismatchError = MintMismatchError;
createErrorFromCodeLookup.set(0x1772, () => new MintMismatchError());
createErrorFromNameLookup.set('MintMismatch', () => new MintMismatchError());
class IndexGreaterThanLengthError extends Error {
    constructor() {
        super('Index greater than length!');
        this.code = 0x1773;
        this.name = 'IndexGreaterThanLength';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, IndexGreaterThanLengthError);
        }
    }
}
exports.IndexGreaterThanLengthError = IndexGreaterThanLengthError;
createErrorFromCodeLookup.set(0x1773, () => new IndexGreaterThanLengthError());
createErrorFromNameLookup.set('IndexGreaterThanLength', () => new IndexGreaterThanLengthError());
class NumericalOverflowErrorError extends Error {
    constructor() {
        super('Numerical overflow error!');
        this.code = 0x1774;
        this.name = 'NumericalOverflowError';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, NumericalOverflowErrorError);
        }
    }
}
exports.NumericalOverflowErrorError = NumericalOverflowErrorError;
createErrorFromCodeLookup.set(0x1774, () => new NumericalOverflowErrorError());
createErrorFromNameLookup.set('NumericalOverflowError', () => new NumericalOverflowErrorError());
class TooManyCreatorsError extends Error {
    constructor() {
        super('Can only provide up to 4 creators to candy machine (because candy machine is one)!');
        this.code = 0x1775;
        this.name = 'TooManyCreators';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, TooManyCreatorsError);
        }
    }
}
exports.TooManyCreatorsError = TooManyCreatorsError;
createErrorFromCodeLookup.set(0x1775, () => new TooManyCreatorsError());
createErrorFromNameLookup.set('TooManyCreators', () => new TooManyCreatorsError());
class UuidMustBeExactly6LengthError extends Error {
    constructor() {
        super('Uuid must be exactly of 6 length');
        this.code = 0x1776;
        this.name = 'UuidMustBeExactly6Length';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, UuidMustBeExactly6LengthError);
        }
    }
}
exports.UuidMustBeExactly6LengthError = UuidMustBeExactly6LengthError;
createErrorFromCodeLookup.set(0x1776, () => new UuidMustBeExactly6LengthError());
createErrorFromNameLookup.set('UuidMustBeExactly6Length', () => new UuidMustBeExactly6LengthError());
class NotEnoughTokensError extends Error {
    constructor() {
        super('Not enough tokens to pay for this minting');
        this.code = 0x1777;
        this.name = 'NotEnoughTokens';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, NotEnoughTokensError);
        }
    }
}
exports.NotEnoughTokensError = NotEnoughTokensError;
createErrorFromCodeLookup.set(0x1777, () => new NotEnoughTokensError());
createErrorFromNameLookup.set('NotEnoughTokens', () => new NotEnoughTokensError());
class NotEnoughSOLError extends Error {
    constructor() {
        super('Not enough SOL to pay for this minting');
        this.code = 0x1778;
        this.name = 'NotEnoughSOL';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, NotEnoughSOLError);
        }
    }
}
exports.NotEnoughSOLError = NotEnoughSOLError;
createErrorFromCodeLookup.set(0x1778, () => new NotEnoughSOLError());
createErrorFromNameLookup.set('NotEnoughSOL', () => new NotEnoughSOLError());
class TokenTransferFailedError extends Error {
    constructor() {
        super('Token transfer failed');
        this.code = 0x1779;
        this.name = 'TokenTransferFailed';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, TokenTransferFailedError);
        }
    }
}
exports.TokenTransferFailedError = TokenTransferFailedError;
createErrorFromCodeLookup.set(0x1779, () => new TokenTransferFailedError());
createErrorFromNameLookup.set('TokenTransferFailed', () => new TokenTransferFailedError());
class CandyMachineEmptyError extends Error {
    constructor() {
        super('Candy machine is empty!');
        this.code = 0x177a;
        this.name = 'CandyMachineEmpty';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, CandyMachineEmptyError);
        }
    }
}
exports.CandyMachineEmptyError = CandyMachineEmptyError;
createErrorFromCodeLookup.set(0x177a, () => new CandyMachineEmptyError());
createErrorFromNameLookup.set('CandyMachineEmpty', () => new CandyMachineEmptyError());
class CandyMachineNotLiveError extends Error {
    constructor() {
        super('Candy machine is not live!');
        this.code = 0x177b;
        this.name = 'CandyMachineNotLive';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, CandyMachineNotLiveError);
        }
    }
}
exports.CandyMachineNotLiveError = CandyMachineNotLiveError;
createErrorFromCodeLookup.set(0x177b, () => new CandyMachineNotLiveError());
createErrorFromNameLookup.set('CandyMachineNotLive', () => new CandyMachineNotLiveError());
class HiddenSettingsConfigsDoNotHaveConfigLinesError extends Error {
    constructor() {
        super('Configs that are using hidden uris do not have config lines, they have a single hash representing hashed order');
        this.code = 0x177c;
        this.name = 'HiddenSettingsConfigsDoNotHaveConfigLines';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, HiddenSettingsConfigsDoNotHaveConfigLinesError);
        }
    }
}
exports.HiddenSettingsConfigsDoNotHaveConfigLinesError = HiddenSettingsConfigsDoNotHaveConfigLinesError;
createErrorFromCodeLookup.set(0x177c, () => new HiddenSettingsConfigsDoNotHaveConfigLinesError());
createErrorFromNameLookup.set('HiddenSettingsConfigsDoNotHaveConfigLines', () => new HiddenSettingsConfigsDoNotHaveConfigLinesError());
class CannotChangeNumberOfLinesError extends Error {
    constructor() {
        super('Cannot change number of lines unless is a hidden config');
        this.code = 0x177d;
        this.name = 'CannotChangeNumberOfLines';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, CannotChangeNumberOfLinesError);
        }
    }
}
exports.CannotChangeNumberOfLinesError = CannotChangeNumberOfLinesError;
createErrorFromCodeLookup.set(0x177d, () => new CannotChangeNumberOfLinesError());
createErrorFromNameLookup.set('CannotChangeNumberOfLines', () => new CannotChangeNumberOfLinesError());
class DerivedKeyInvalidError extends Error {
    constructor() {
        super('Derived key invalid');
        this.code = 0x177e;
        this.name = 'DerivedKeyInvalid';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, DerivedKeyInvalidError);
        }
    }
}
exports.DerivedKeyInvalidError = DerivedKeyInvalidError;
createErrorFromCodeLookup.set(0x177e, () => new DerivedKeyInvalidError());
createErrorFromNameLookup.set('DerivedKeyInvalid', () => new DerivedKeyInvalidError());
class PublicKeyMismatchError extends Error {
    constructor() {
        super('Public key mismatch');
        this.code = 0x177f;
        this.name = 'PublicKeyMismatch';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, PublicKeyMismatchError);
        }
    }
}
exports.PublicKeyMismatchError = PublicKeyMismatchError;
createErrorFromCodeLookup.set(0x177f, () => new PublicKeyMismatchError());
createErrorFromNameLookup.set('PublicKeyMismatch', () => new PublicKeyMismatchError());
class NoWhitelistTokenError extends Error {
    constructor() {
        super('No whitelist token present');
        this.code = 0x1780;
        this.name = 'NoWhitelistToken';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, NoWhitelistTokenError);
        }
    }
}
exports.NoWhitelistTokenError = NoWhitelistTokenError;
createErrorFromCodeLookup.set(0x1780, () => new NoWhitelistTokenError());
createErrorFromNameLookup.set('NoWhitelistToken', () => new NoWhitelistTokenError());
class TokenBurnFailedError extends Error {
    constructor() {
        super('Token burn failed');
        this.code = 0x1781;
        this.name = 'TokenBurnFailed';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, TokenBurnFailedError);
        }
    }
}
exports.TokenBurnFailedError = TokenBurnFailedError;
createErrorFromCodeLookup.set(0x1781, () => new TokenBurnFailedError());
createErrorFromNameLookup.set('TokenBurnFailed', () => new TokenBurnFailedError());
class GatewayAppMissingError extends Error {
    constructor() {
        super('Missing gateway app when required');
        this.code = 0x1782;
        this.name = 'GatewayAppMissing';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, GatewayAppMissingError);
        }
    }
}
exports.GatewayAppMissingError = GatewayAppMissingError;
createErrorFromCodeLookup.set(0x1782, () => new GatewayAppMissingError());
createErrorFromNameLookup.set('GatewayAppMissing', () => new GatewayAppMissingError());
class GatewayTokenMissingError extends Error {
    constructor() {
        super('Missing gateway token when required');
        this.code = 0x1783;
        this.name = 'GatewayTokenMissing';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, GatewayTokenMissingError);
        }
    }
}
exports.GatewayTokenMissingError = GatewayTokenMissingError;
createErrorFromCodeLookup.set(0x1783, () => new GatewayTokenMissingError());
createErrorFromNameLookup.set('GatewayTokenMissing', () => new GatewayTokenMissingError());
class GatewayTokenExpireTimeInvalidError extends Error {
    constructor() {
        super('Invalid gateway token expire time');
        this.code = 0x1784;
        this.name = 'GatewayTokenExpireTimeInvalid';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, GatewayTokenExpireTimeInvalidError);
        }
    }
}
exports.GatewayTokenExpireTimeInvalidError = GatewayTokenExpireTimeInvalidError;
createErrorFromCodeLookup.set(0x1784, () => new GatewayTokenExpireTimeInvalidError());
createErrorFromNameLookup.set('GatewayTokenExpireTimeInvalid', () => new GatewayTokenExpireTimeInvalidError());
class NetworkExpireFeatureMissingError extends Error {
    constructor() {
        super('Missing gateway network expire feature when required');
        this.code = 0x1785;
        this.name = 'NetworkExpireFeatureMissing';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, NetworkExpireFeatureMissingError);
        }
    }
}
exports.NetworkExpireFeatureMissingError = NetworkExpireFeatureMissingError;
createErrorFromCodeLookup.set(0x1785, () => new NetworkExpireFeatureMissingError());
createErrorFromNameLookup.set('NetworkExpireFeatureMissing', () => new NetworkExpireFeatureMissingError());
class CannotFindUsableConfigLineError extends Error {
    constructor() {
        super('Unable to find an unused config line near your random number index');
        this.code = 0x1786;
        this.name = 'CannotFindUsableConfigLine';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, CannotFindUsableConfigLineError);
        }
    }
}
exports.CannotFindUsableConfigLineError = CannotFindUsableConfigLineError;
createErrorFromCodeLookup.set(0x1786, () => new CannotFindUsableConfigLineError());
createErrorFromNameLookup.set('CannotFindUsableConfigLine', () => new CannotFindUsableConfigLineError());
class InvalidStringError extends Error {
    constructor() {
        super('Invalid string');
        this.code = 0x1787;
        this.name = 'InvalidString';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, InvalidStringError);
        }
    }
}
exports.InvalidStringError = InvalidStringError;
createErrorFromCodeLookup.set(0x1787, () => new InvalidStringError());
createErrorFromNameLookup.set('InvalidString', () => new InvalidStringError());
class SuspiciousTransactionError extends Error {
    constructor() {
        super('Suspicious transaction detected');
        this.code = 0x1788;
        this.name = 'SuspiciousTransaction';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, SuspiciousTransactionError);
        }
    }
}
exports.SuspiciousTransactionError = SuspiciousTransactionError;
createErrorFromCodeLookup.set(0x1788, () => new SuspiciousTransactionError());
createErrorFromNameLookup.set('SuspiciousTransaction', () => new SuspiciousTransactionError());
class CannotSwitchToHiddenSettingsError extends Error {
    constructor() {
        super('Cannot Switch to Hidden Settings after items available is greater than 0');
        this.code = 0x1789;
        this.name = 'CannotSwitchToHiddenSettings';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, CannotSwitchToHiddenSettingsError);
        }
    }
}
exports.CannotSwitchToHiddenSettingsError = CannotSwitchToHiddenSettingsError;
createErrorFromCodeLookup.set(0x1789, () => new CannotSwitchToHiddenSettingsError());
createErrorFromNameLookup.set('CannotSwitchToHiddenSettings', () => new CannotSwitchToHiddenSettingsError());
class IncorrectSlotHashesPubkeyError extends Error {
    constructor() {
        super('Incorrect SlotHashes PubKey');
        this.code = 0x178a;
        this.name = 'IncorrectSlotHashesPubkey';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, IncorrectSlotHashesPubkeyError);
        }
    }
}
exports.IncorrectSlotHashesPubkeyError = IncorrectSlotHashesPubkeyError;
createErrorFromCodeLookup.set(0x178a, () => new IncorrectSlotHashesPubkeyError());
createErrorFromNameLookup.set('IncorrectSlotHashesPubkey', () => new IncorrectSlotHashesPubkeyError());
class IncorrectCollectionAuthorityError extends Error {
    constructor() {
        super('Incorrect collection NFT authority');
        this.code = 0x178b;
        this.name = 'IncorrectCollectionAuthority';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, IncorrectCollectionAuthorityError);
        }
    }
}
exports.IncorrectCollectionAuthorityError = IncorrectCollectionAuthorityError;
createErrorFromCodeLookup.set(0x178b, () => new IncorrectCollectionAuthorityError());
createErrorFromNameLookup.set('IncorrectCollectionAuthority', () => new IncorrectCollectionAuthorityError());
class MismatchedCollectionPDAError extends Error {
    constructor() {
        super('Collection PDA address is invalid');
        this.code = 0x178c;
        this.name = 'MismatchedCollectionPDA';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, MismatchedCollectionPDAError);
        }
    }
}
exports.MismatchedCollectionPDAError = MismatchedCollectionPDAError;
createErrorFromCodeLookup.set(0x178c, () => new MismatchedCollectionPDAError());
createErrorFromNameLookup.set('MismatchedCollectionPDA', () => new MismatchedCollectionPDAError());
class MismatchedCollectionMintError extends Error {
    constructor() {
        super("Provided mint account doesn't match collection PDA mint");
        this.code = 0x178d;
        this.name = 'MismatchedCollectionMint';
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, MismatchedCollectionMintError);
        }
    }
}
exports.MismatchedCollectionMintError = MismatchedCollectionMintError;
createErrorFromCodeLookup.set(0x178d, () => new MismatchedCollectionMintError());
createErrorFromNameLookup.set('MismatchedCollectionMint', () => new MismatchedCollectionMintError());
function errorFromCode(code) {
    const createError = createErrorFromCodeLookup.get(code);
    return createError != null ? createError() : null;
}
exports.errorFromCode = errorFromCode;
function errorFromName(name) {
    const createError = createErrorFromNameLookup.get(name);
    return createError != null ? createError() : null;
}
exports.errorFromName = errorFromName;
